src/controllers/controller.bar.js

```diff
@@ -243,17 +243,31 @@ module.exports = function(Chart) {
 			var stackCount = me.getStackCount();
 			var datasetIndex = me.index;
 			var pixels = [];
+			var orderedPixels = [];
+			var order = [];
 			var isHorizontal = scale.isHorizontal();
 			var start = isHorizontal ? scale.left : scale.top;
 			var end = start + (isHorizontal ? scale.width : scale.height);
-			var i, ilen;
+			var i, ilen, pixel;
 
 			for (i = 0, ilen = me.getMeta().data.length; i < ilen; ++i) {
-				pixels.push(scale.getPixelForValue(null, i, datasetIndex));
+				pixel = scale.getPixelForValue(null, i, datasetIndex);
+				if (pixel !== undefined) {
+					pixels.push({index: i, value: pixel});
+				}
+			}
+
+			pixels.sort(function(a, b) {
+				return a.value - b.value;
+			});
+			for (i = 0, ilen = pixels.length; i < ilen; ++i) {
+				orderedPixels.push(pixels[i].value);
+				order[pixels[i].index] = i;
```

Comment:
I don't think we need to adjust the data arrays in datasets because the `order` array translates an original index into an ordered index, and the ordered indexes and pixels are only used for bar width calculation. Do you have any specific problematic cases in mind?